source/board/microbitv2/microbitv2.c

```diff
@@ -18,23 +19,591 @@
+          } else if (shutdown_led_dc == 10) {
+              shutdown_led_dc = PWR_LED_ON_MAX_BRIGHTNESS;
+          } else if (shutdown_led_dc <= (PWR_LED_ON_MAX_BRIGHTNESS - 10)) {
+              shutdown_led_dc = 0;
+          } else if (shutdown_led_dc > (PWR_LED_ON_MAX_BRIGHTNESS - 10)) {
+              shutdown_led_dc--;
+          }
+          
+          // If we are no longer PC connected, go into idle mode
+          if (usb_state == USB_DISCONNECTED && !usb_pc_connected) {
+              main_shutdown_state = MAIN_SHUTDOWN_WAITING;
+          }
+          break;
+      case MAIN_LED_BLINK_ONCE:
+          
+          if (blink_in_progress) {
+            blink_in_progress--;
+            if (blink_in_progress == 5) {
+              shutdown_led_dc = 0;
+            }
+          } else {
+            blink_in_progress = 10;
+            shutdown_led_dc = PWR_LED_ON_MAX_BRIGHTNESS;
+          }
+          
+          if (blink_in_progress == 0) { 
+            main_shutdown_state = MAIN_SHUTDOWN_WAITING;
+          }
+          break;
+      default:
+          break;
+    }
+
+    // Use gamma curve except in final fade
+    if (main_shutdown_state != MAIN_SHUTDOWN_REACHED_FADE) {
+        uint8_t gamma_led_dc = get_led_gamma(shutdown_led_dc);
+        flexio_pwm_set_dutycycle(gamma_led_dc);
+    } else {
+        flexio_pwm_set_dutycycle(final_fade_led_dc);
+    }
+    
+    // Remount if requested.
+    if (do_remount) {
+        do_remount = false;
+        vfs_mngr_fs_remount();
+    }
+}
+
+void board_handle_powerdown()
+{
+    switch(interface_power_mode){
+        case kAPP_PowerModeVlps:
+            power_enter_VLPS();
+            break;
+        case kAPP_PowerModeVlls0:
+            power_enter_VLLS0();
+            break;
+        default:
+            break;
+    }
+}
+
+void board_usb_sof_event(void)
+{
+    /* Set usb_pc_connected to true to prevent entering sleep mode when host does re-enumeration.
+    *  Will be set to false when the USB cable is detached. */
+    usb_pc_connected = true;
+}
+
+void board_vfs_stream_closed_hook()
+{
+    // Return reset button and LED to default state
+    reset_pressed = 0;
+    power_led_sleep_state_on = PWR_LED_SLEEP_STATE_DEFAULT;
+    main_shutdown_state = MAIN_SHUTDOWN_WAITING;
+    
+    // Clear any pending I2C response
+    i2c_clearBuffer();
+    
+    // Release COMBINED_SENSOR_INT
+    PORT_SetPinMux(COMBINED_SENSOR_INT_PORT, COMBINED_SENSOR_INT_PIN, kPORT_PinDisabledOrAnalog);
+}
+
+bool vfs_user_magic_file_hook(const vfs_filename_t filename, bool *do_remount)
+{
+    if (!memcmp(filename, "ERASE   ACT", sizeof(vfs_filename_t))) {
+        // Erase last 128KB flash block
+        FLASH_Erase(&g_flash, FLASH_CONFIG_ADDRESS, g_flash.PFlashTotalSize / g_flash.PFlashBlockCount, kFLASH_apiEraseKey);
+    }
+    
+    return false;
 }
 
-// USB HID override function return 1 if the activity is trivial or response is null 
+void vfs_user_build_filesystem_hook() {
```

Comment:
When DAPLink encounters an irrecoverable error we erase the target flash and flash a small program on the nRF to scroll the error code on the display.

The error codes come from an enum in `source/daplink/error.h`: https://github.com/microbit-foundation/DAPLink-microbit/blob/d7f24a673dc417fd169bd1f2ae196eb5a796303b/source/daplink/error.h#L30-L85

The enum is grouped in categories, so if future DAPLink updates add extra errors in one of the categories before the end of the enum, then all subsequent error values will shift.

It's quite important for us that the values shown to the user don't change, to be able to maintain consistency and so that online documentation of the error codes is always correct.
I guess the only way to "fix in place" the error values would be to only add new errors at the end of the enum. Is this something you would consider? Otherwise we might need to convert these values here, but how that is done would depend on the changes in `error_t` and it could be easy to accidentally update one without the other.

Reply:
Added comment at the top of the enum in error.h.